File: ddz_admin/app/models/task_def.rb
class TaskDef
  include Mongoid::Document

  field :taskId, type: String
  field :taskName, type: String
  field :taskDesc, type: String
  field :taskType, type: String
  field :taskIcon, type: String
  field :taskBonusDesc, type: String
  field :sortIndex, type: Integer, default: 255
  field :enabled, type: Boolean, default: true
  field :progress, type: Integer, default: 0
  field :progressDesc, type: String
  field :taskData, type: Hash, default: ->{ {:_placeholder => 0} }
  include Mongoid::Timestamps
  # field :createdAt, type: Date, default: ->{ Time.now }
  # field :updatedAt, type: Date, default: ->{ Time.now }

  validates :taskId, uniqueness: true, presence: true
  validates :taskName, uniqueness: true, presence: true


  def self.serialize_from_session(key, salt)
    record = to_adapter.get((key[0]["$oid"] rescue nil))
    record if record && record.authenticatable_salt == salt
  end


  def taskData_hash
    self.taskData.to_json
  end

  def taskData_hash=(value)
    self.taskData = JSON.parse(value)
  end
end

File: charge-server/app/domain/taskDef.js
/**
 * Copyright (c) 2015 深圳市辉游科技有限公司.
 */

var mongoose = require('mongoose-q')();
var Schema = mongoose.Schema;
var uuid = require('node-uuid');
var utils = require('../util/utils');
var DomainUtils = require("./domainUtils");

/**
 * 任务定义
 * @type {Mongoose.Schema}
 */
var taskDefSchema = new mongoose.Schema({
  taskId: {type: String, required: "{task id} is required!"},
  taskName: {type: String, required: "{task name} is required!"},
  taskDesc: {type: String, required: "{task description} is required!"},
  taskType: {type: String, required: true},
  taskIcon: {type: String},
  taskBonusDesc: {type: String},
  sortIndex: {type: Number, default: 255},
  enabled: {type: Boolean, default: true},
  progress: {type: Number, default: 0},
  progressDesc: String,
  taskProcessor: String,
  taskTrigger: String,
  taskData: {type: Schema.Types.Mixed, default: {_placeholder:0}},
  created_at: {type: Date, default: Date.now},
  updated_at: {type: Date, default: Date.now}
}, {
  collection: 'task_defs'
});


var __toParams = function(model, opts) {
  var transObj = {
    userId: 0,
    taskId: model.taskId,
    taskName: model.taskName,
    taskDesc: model.taskDesc,
    taskIcon: model.taskIcon,
    taskType: model.taskType,
    taskBonusDesc: model.taskBonusDesc,
    progress: model.progress,
    progressDesc: model.progressDesc
  };

  transObj = DomainUtils.adjustAttributes(transObj, opts);

  return transObj;
};

taskDefSchema.statics.toParams = __toParams;

taskDefSchema.methods.toParams = function(opts) {
  return __toParams(this, opts);
};


var TaskDef = mongoose.model('TaskDef', taskDefSchema);

module.exports = TaskDef;